### PR TITLE
blog: libmemkind intro

### PR DESCRIPTION
**Pull request by jschmieg**
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/118)

### CODE REVIEW

File: _posts/2020-01-20-libmemkind.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+title: Introduction to libmemkind
+author: jschmieg
+layout: post
+identifier: libmemkind
+---
+
+
+# Introduction
+
+One of the components of PMDK is libmemkind, a library which was developed to
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: 1) memkind is not part of PMDK, no matter how crazy logic you will apply
2) memkind was not created for persistent memory, it was updated to support it

File: _posts/2020-01-20-libmemkind.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+simplify
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: what's up with the formatting?

File: _posts/2020-01-20-libmemkind.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+allows partitioning a heap of an application between these kinds. On a system
+equipped
+with DRAM and NVDIMMs, it is possible to re-design an application in a way 
+to store objects that are accessed frequently and require fast access in DRAM
+while larger objects which are accessed less frequently can be stored on
+persistent memory.
+
+# Managing application heap
+
+With libmemkind it is possible to allocate in DRAM by calling `memkind_malloc`
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: from DRAM

File: _posts/2020-01-20-libmemkind.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+function with static kind `MEMKIND_DEFAULT` or `MEMKIND_REGULAR`:
+```c
+void * ptr_default = memkind_malloc(MEMKIND_DEFAULT, size);
+```
+and in persistent memory by calling the same function with dynamically created
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: from

File: _posts/2020-01-20-libmemkind.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+pmem kind pointing to directory on FS DAX:
+```c
+struct memkind *pmem_kind = NULL;
+memkind_create_pmem("/mnt/pmem", PMEM_MAX_SIZE, &pmem_kind);
+void * ptr_pmem = memkind_malloc(pmem_kind, size);
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: space between * and variable name should not be there

File: _posts/2020-01-20-libmemkind.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+Freeing allocated objects is done by calling:
+```c
+memkind_free(MEMKIND_DEFAULT, ptr_default);
+memkind_free(pmem_kind, ptr_pmem);
+```
+To simplify the adoption of applications to many memory pools, it is worth using
+another library feature: calling `free` without specifying a kind that was used
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: memkind_free,

this whole sentence sounds bad

File: _posts/2020-01-20-libmemkind.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+for the allocation. Library will automatically recognize if a given pointer
+belongs to DRAM or PM. This precludes tracking each allocation origin in the
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: avoids tracking of each allocation's origin

File: _posts/2020-01-20-libmemkind.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+identifier: libmemkind
+---
+
+
+# Introduction
+
+One of the components of PMDK is libmemkind, a library which was developed to
+simplify
+usage of persistent memory in a volatile mode. There are NVDIMMs technologies,
+such as Intel Optane DCPMM, that provides persistency, byte-addressability, but
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: "but"?

File: _posts/2020-01-20-libmemkind.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+
+# Introduction
+
+One of the components of PMDK is libmemkind, a library which was developed to
+simplify
+usage of persistent memory in a volatile mode. There are NVDIMMs technologies,
+such as Intel Optane DCPMM, that provides persistency, byte-addressability, but 
+also a high capacity when compared with DRAM modules. By extension, they can be 
+used as an augmentation of main memory and utilized by applications which
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: By extension -> /dev/null?
augmentation -> expansion?

File: _posts/2020-01-20-libmemkind.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+databases, 
+caching engines and scientific simulations.
+
+# Libmemkind
+
+According to the current persistent memory programming model, NVDIMMs are
+exposed by operating system as devices on which user should create file-system.
+This model creates a need for a way to consume memory exposed through files in 
+applications. Libmemkind fills this gap 
+by utilizing jemalloc on temporary files created on NVDIMMs and acts as a memory
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: files are created on a file system, not on the dimm directly

File: _posts/2020-01-20-libmemkind.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+allocator for applications. The new option of utilizing persistent memory as a
+hotplugged memory region (KMEM DAX) is already available in Linux kernel
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: The information about KMEM DAX doesn't belong here.

File: _posts/2020-01-20-libmemkind.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+beginning from
+version 5.1 and Memkind support for this feature will be described in the next
+blog entry.
+Libmemkind provides various memory pools called “kinds” for memories with
+miscellaneous characteristics: DRAM, persistent memory and High-Bandwidth
+Memory. This
+allows partitioning a heap of an application between these kinds. On a system
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: "partitioning a heap of an application" doesn't make any sense

File: _posts/2020-01-20-libmemkind.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+equipped
+with DRAM and NVDIMMs, it is possible to re-design an application in a way
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: redesign is pretty heavy word for such change...

File: _posts/2020-01-20-libmemkind.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+for the allocation. Library will automatically recognize if a given pointer
+belongs to DRAM or PM. This precludes tracking each allocation origin in the
+application:
+```c
+memkind_free(NULL, ptr_default);
+memkind_free(NULL, ptr_pmem);
+```
+
+# Other features
+Libmemkind provides also:
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: also provides

File: _posts/2020-01-20-libmemkind.md [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+* rest of malloc-style API functions,
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: the rest of

File: _posts/2020-01-20-libmemkind.md [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+```c
+memkind_free(NULL, ptr_default);
+memkind_free(NULL, ptr_pmem);
+```
+
+# Other features
+Libmemkind provides also:
+* rest of malloc-style API functions, 
+* an automatically increased pool size,
+* configuring memory policy: performance vs fragmentation,
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: configurable

File: _posts/2020-01-20-libmemkind.md [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+* C++ bindings,
+* choosing alternative heap manager (TBB).
```
Comment (marcinslusarz) [inline _posts/2020-01-20-libmemkind.md]: ~choosing~

File: _posts/2020-01-20-libmemkind.md (hunk 1 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 2 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 8 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 9 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 10 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 11 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: Done, however, I was asked to include such sentence in previous reviews.

File: _posts/2020-01-20-libmemkind.md (hunk 12 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: I did small correction, however description is similar to Memkind main page tittle:
http://memkind.github.io/memkind/

File: _posts/2020-01-20-libmemkind.md (hunk 13 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 3 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 4 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 5 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 6 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 7 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 14 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 15 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 16 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

File: _posts/2020-01-20-libmemkind.md (hunk 17 above)
Comment (jschmieg) [reply _posts/2020-01-20-libmemkind.md]: done

**Pull request merged**